node/gum/proc-macro/src/lib.rs

```diff
@@ -0,0 +1,171 @@
+	gum(item, Level::Debug)
+}
+
+/// Print a trace level message.
+#[proc_macro]
+pub fn trace(item: proc_macro::TokenStream) -> proc_macro::TokenStream {
+	gum(item, Level::Trace)
+}
+
+/// One-size-fits all internal implementation that produces the actual code.
+pub(crate) fn gum(item: proc_macro::TokenStream, level: Level) -> proc_macro::TokenStream {
+	let item: TokenStream = item.into();
+
+	let res = expander::Expander::new("gum")
+		.add_comment("Generated overseer code by `gum::warn!(..)`".to_owned())
+		.dry(!cfg!(feature = "expand"))
+		.verbose(false)
+		.fmt(expander::Edition::_2021)
+		.maybe_write_to_out_dir(impl_gum2(item, level))
+		.expect("Expander does not fail due to IO in OUT_DIR. qed");
+
+	res.unwrap_or_else(|err| err.to_compile_error()).into()
+}
+
+/// Does the actual parsing and token generation based on `proc_macro2` types.
+///
+/// Required for unit tests.
+pub(crate) fn impl_gum2(orig: TokenStream, level: Level) -> Result<TokenStream> {
+	let args: Args = parse2(orig)?;
+
+	let krate = support_crate();
+	let span = Span::call_site();
+
+	let Args { target, comma, mut values, fmt } = args;
+
+	// find a value or alias called `candidate_hash`.
+	let maybe_candidate_hash = values.iter_mut().find(|value| value.as_ident() == "candidate_hash");
+
+	if let Some(kv) = maybe_candidate_hash {
+		let (ident, rhs_expr, replace_with) = match kv {
+			Value::Alias(alias) => {
+				let ValueWithAliasIdent { alias, marker, expr, .. } = alias.clone();
+				(
+					alias.clone(),
+					expr.to_token_stream(),
+					Some(Value::Value(ValueWithFormatMarker {
+						marker,
+						ident: alias,
+						dot: None,
+						inner: Punctuated::new(),
+					})),
+				)
+			},
+			Value::Value(value) => (value.ident.clone(), value.ident.to_token_stream(), None),
+		};
+
+		// we generate a local value with the same alias name
+		// so replace the expr with just a value
+		if let Some(replace_with) = replace_with {
+			let _old = std::mem::replace(kv, replace_with);
+		};
+
+		// Inject the addition `traceID = % trace_id` identifier
+		// while maintaining trailing comma semantics.
+		let had_trailing_comma = values.trailing_punct();
+		if !had_trailing_comma {
+			values.push_punct(Token![,](span));
+		}
+
+		values.push_value(parse_quote! {
+			traceID = % trace_id
+		});
+		if had_trailing_comma {
+			values.push_punct(Token![,](span));
+		}
+
+		Ok(quote! {
+			if #krate :: enabled!(#target #comma #level) {
+				use ::std::ops::Deref;
+
+				// create a scoped let binding of something that `deref`s to
+				// `Hash`.
+				let value = #rhs_expr;
+				let value = &value;
+				let value: & #krate:: Hash = value.deref();
+				// Do the `deref` to `Hash` and convert to a `TraceIdentifier`.
+				let #ident: #krate:: Hash = * value;
+				let trace_id = #krate:: hash_to_trace_identifier ( #ident );
+				#krate :: event!(
+					#target #comma #level, #values #fmt
+				)
+			}
+		})
+	} else {
+		Ok(quote! {
```

Comment:
Would a falling back to `relay_parent` add value here ?

Reply:
I think `relay_parent` might be another tracing id in conjunction with `candidate_hash` as it is available in more places than `candidate_hash`. But I also think that these IDs should be kept separately, as fallback would only mess things up.

Reply:
Yeah, that would mess things up, having a secondary trace id is what makes sense here. Falling back to a secondary trace id is what I really meant here.